docs/labels.mdx

```diff
@@ -30,3 +30,19 @@ render(
   </div>
 )
 ```
+
+## SSR & Safari
```

Comment:
I wonder if there is any way to opt-out of this optimization in our dev build. I don't see how this would be possible as it is applied to a function (a component) that is not under our control.

The optimization kicks in when the returned element from a component is the CSS prop element. Like here:
```jsx
function Home() {
  return <div css={{ color: 'orchid' }}>Some colored text</div>
}
```

It is enough to not return this element directly to prevent this optimization from kicking in:
```jsx
function Home() {
  const el = <div css={{ color: 'orchid' }}>Some colored text</div>
  return el
}
```
but as mentioned - the `Home` function in this example is out of our control so I'm not sure how we possibly could prevent it from happening 🤔 

Maybe we should reach for [`suppressHydrationWarning`](https://reactjs.org/docs/dom-elements.html#suppresshydrationwarning) in this situation?

This creating a confusing warning for the user in one of the browsers is IMHO really bad - to the point that I don't think that having this logic with the warning is worth it.

@mitchellhamilton any thoughts on the subject? what do you think about dropping this runtime extraction or using `suppressHydrationWarning` in Safari?

Reply:
We could also try to minimize the need for the suppression prop by consulting the DOM, some labels could be extracted from the SSRed HTML

Reply:
Well, the suppression flag is a dumb idea as it would most likely hide other, legit, hydration mismatches that could happen there

Reply:
> We could also try to minimize the need for the suppression prop by consulting the DOM, some labels could be extracted from the SSRed HTML

Very interesting idea. It seems like it would really hard to find the correct DOM node, but if we could do that the label extraction would be simple.

The only other options I can think of are:
- A global flag to either opt-in or opt-out of label extraction.
- Remove label extraction period. I don't like this since the labels are very useful.

But I defer to you and Mitchell.

Reply:
> suppressHydrationWarning

yeah, i'm not really into doing that

> We could also try to minimize the need for the suppression prop by consulting the DOM, some labels could be extracted from the SSRed HTML

How are you thinking this could work?

> what do you think about dropping this runtime extraction

tbh, I've been kinda wanting to do this because I've got a project where I render a _lot_ of elements and the label extraction slows it down to the point where it's faster for me to do a production build and use that than using the dev version(with the label extraction, it takes ~1.2 minutes vs ~10 seconds with the label extraction removed from the package but still in dev to hydrate the tree and basically all that cost is the regex). If we do want to keep it, we should really improve the performance. (Here are traces with and without the label extraction if you're curious, you'll need to unzip them before loading them into Chrome devtools - [with-label-extraction.json.zip](https://github.com/emotion-js/emotion/files/7535028/with-label-extraction.json.zip), [without-label-extraction.json.zip](https://github.com/emotion-js/emotion/files/7535030/without-label-extraction.json.zip))

Reply:
> How are you thinking this could work?

The problem with this solution is that it won't work perfectly for:
```js
const style = css`color: hotpink;`

function Comp1() { return <div css={style} /> }
function Comp2() { return <div css={style} /> }
```

It's because this now results in 2 different labels - but in just a single hash. This doesn't sound like a super common situation though.

Since the whole thing in Safari can only affect SSR scenarios we should be able to extract the label from the data attribute on `document.querySelector('[data-emotion*="' + hash + '"]').dataset.emotion`. 

At the moment the regexp parsing happens in the `createEmotionProps`, while serializing styles happen in the Emotion element - so this thing I'm talking about would have to be added to the latter. 

Unsure if this is a good idea - just an idea I've had.

> If we do want to keep it, we should really improve the performance. (Here are traces with and without the label extraction if you're curious, you'll need to unzip them before loading them into Chrome devtools - with-label-extraction.json.zip, without-label-extraction.json.zip)

could you share actual stack traces? it's not totally straightforward to grab them from the profiles - or I don't know how to do that. I see though that this takes a lot of time for u there! 😱 

@srmagura has refactored this extraction quite a bit here and it has potential of being fast, we could also try to optimize this even further. I think it's worth to give this a shot - even to learn how fast we can get with that and how much of an impact this kind of a feature can have on the runtime. Would you be willing to try out a snapshot release in your project?

Reply:
I'm happy to benchmark the new implementation against Mitchell's stack traces, but I think we should first decide if label extraction is being removed. No point in optimizing code if we end up deleting it.

That said, I think the new stack trace parsing should already be pretty fast, with the potential for further optimization.

Reply:
I tried this branch on my project and it totally removed the regex being visible from the trace so that's great and it's  down to within a couple seconds of the version without the label extraction.

Also, applying the patch below and adding labels to some things practically completely removed the difference between the version with and without label extraction given it's basically never running.

```patch
diff --git a/packages/react/src/emotion-element.js b/packages/react/src/emotion-element.js
index d2f97cd1..df5f689b 100644
--- a/packages/react/src/emotion-element.js
+++ b/packages/react/src/emotion-element.js
@@ -33,7 +33,12 @@ export const createEmotionProps = (type: React.ElementType, props: Object) => {
 
   newProps[typePropName] = type
 
-  if (process.env.NODE_ENV !== 'production') {
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    (typeof props.css !== 'object' ||
+      typeof props.css.name !== 'string' ||
+      props.css.name.indexOf('-') === -1)
+  ) {
     const label = getLabelFromStackTrace(new Error().stack)
     if (label) newProps[labelPropName] = label
   }
```

> we should first decide if label extraction is being removed

I think the problems with it are:

- Performance
- The Safari thing

Given the changes already in this PR, the performance impact is now pretty minimal and given this is in dev, I think a small performance impact is totally fine.
I'm not that concerned about the Safari issue given it's just a warning.
Also, given the patch above + using the Babel plugin, the performance impact doesn't really exist since you'll basically always skip that code and you won't get problems on Safari given the labels are added by the Babel plugin.

So I think I'm okay with leaving it in unless you disagree @Andarist?

Reply:
An idea for addressing the Safari issue: what if we overwrite `console.error` and don't call the real `console.error` warnings when we know they're fine?

Ideally we would even look at the content of the message, get the client side class name, look in the cache for the content for that, add `label:LabelExtractedFromServerSideClassName;`, hash it and compare the names and only skip it when they match so you'd still get the warning when it's correct

Reply:
I put in your performance improvement @mitchellhamilton! (with an added null check)

I like the `console.error` idea and can try that out on a new branch. Am I right in thinking that I will have to overwrite `console.error` once per Emotion cache? Since there could be multiple caches used on a single page.

Reply:
> Am I right in thinking that I will have to overwrite console.error once per Emotion cache?

Yes

Reply:
> Also, applying the patch below and adding labels to some things practically completely removed the difference between the version with and without label extraction given it's basically never running.

Could you confirm that the performance improvement was still huge without this patch? From what I read from it - it checks if the value passed to the css prop has been transformed by our Babel plugin, right? Since there is, indeed, no need to get the label from the stack trace in such a case - this is a good optimization for Babel users. I'd like to make sure that the performance improvements prepared by @srmagura are enough for non-Babel users. I'm pretty sure that they are - just double-checking.

> So I think I'm okay with leaving it in unless you disagree @Andarist?

Assuming the perf of this is fine then I'm OK with leaving this as is - it's a useful feature. Somebody has suggested in the past that sometimes mere `new Error().stack` can be a major perf hit but I was never able to experience that. No profile dumps that I've investigated showed that.

> An idea for addressing the Safari issue: what if we overwrite console.error and don't call the real console.error warnings when we know they're fine?

This has crossed my mind but I've concluded that something like this is always risky - although probably would work in a vast majority of cases. I would have to recheck but I think that, for instance, React DevTools might also overwrite those. I don't see how we could, for example, clash with their logic but if we could stay away from that - it would be IMHO preferable. 

What do you think about the querySelector approach that would attempt to minimize the mismatch? With proper caching and stuff it feels like it could take us quite far and **without** losing the label - which is tempting. Given that you propose rehashing stuff from within the overridden `console.error` etc I'm not sure if the `console.error` approach would end up being much simpler, implementation-wise. An argument in favor of the `console.error` approach would definitely be that we would never report any "false positives" to users - although with the querySelector approach we potentially could assess if the mismatch can happen (in rare, rare cases) and print a helpful information to the console that one should ignore the React warning and that this can be "fixed" by using a Babel plugin.

Reply:
btw. nice teamwork everybody! 

<img src="https://media.giphy.com/media/xUPGcjQ6dJEjH5uwMw/giphy-downsized-large.gif" />

Reply:
Love the teamwork 😎. 

There are definitely other tools that overwrite `console.error` e.g. error monitoring libraries like Bugsnag. The risk I see is that the "class name mismatch" error could get logged somewhere before we have a chance to intercept it.

The querySelector approach seems simpler but, as you said, it does not handle this case:

```ts
const style = css`color: hotpink;`

function Comp1() { return <div css={style} /> }
function Comp2() { return <div css={style} /> }
```

I'm also not sure how common this is, but it is a valid use case. Maybe you have hotpink text all over your site and want to just import that same `style` object everywhere. We would end up giving all these elements the same label which would be confusing.

I think `console.error` would be my preference.

BTW, could we merge this now and I'll do a separate PR for working around the Safari issue? Since I believe the new `getLabelByStackTrace` is an improvement over what's in `main` despite it not solving the Safari issue.

Reply:
> I'm also not sure how common this is, but it is a valid use case. Maybe you have hotpink text all over your site and want to just import that same style object everywhere. We would end up giving all these elements the same label which would be confusing.

Importing the same style object is very much a common practice - I just assume that in such situations such an object is mixed together with some other styles. Since we "flatten" all the styles into a single class name the resulting hash should be different and thus maybe this would work almost perfectly in practice. Hard to tell though as we don't have any good data on how people are writing their styles in the real world.

Reply:
> Could you confirm that the performance improvement was still huge without this patch?

yes, it was

> What do you think about the querySelector approach that would attempt to minimize the mismatch?

I don't really see how it would work, even ignoring styles being used with different labels, given that the hash is based on the style with the `label:Blah;` how would we match a style to a style with a label?

> This has crossed my mind but I've concluded that something like this is always risky - although probably would work in a vast majority of cases. I would have to recheck but I think that, for instance, React DevTools might also overwrite those. I don't see how we could, for example, clash with their logic but if we could stay away from that - it would be IMHO preferable.

React DevTools does override console.error but yeah, I don't see how it would clash. Maybe depending on which one overrides first, that could break things?

Reply:
> I don't really see how it would work, even ignoring styles being used with different labels, given that the hash is based on the style with the label:Blah; how would we match a style to a style with a label?

🤦 I forgot that the label is part of the hashed content. 

> React DevTools does override console.error but yeah, I don't see how it would clash. Maybe depending on which one overrides first, that could break things?

Ye, maybe - hard to assess this without actually trying this out